modules/core/src/test/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotTraverseSuite.scala

```diff
@@ -0,0 +1,266 @@
+package org.tessellation.dag.snapshot
+
+import java.security.KeyPair
+
+import cats.data.{NonEmptyList, NonEmptySet}
+import cats.effect.std.Random
+import cats.effect.unsafe.implicits.global
+import cats.effect.{IO, Resource}
+import cats.syntax.applicative._
+import cats.syntax.eq._
+import cats.syntax.foldable._
+import cats.syntax.traverse._
+
+import scala.collection.immutable.{SortedMap, SortedSet}
+
+import org.tessellation.domain.rewards.Rewards
+import org.tessellation.domain.snapshot.SnapshotStorage
+import org.tessellation.domain.statechannel.StateChannelValidator
+import org.tessellation.ext.cats.effect.ResourceIO
+import org.tessellation.ext.cats.syntax.next._
+import org.tessellation.infrastructure.snapshot.{
+  GlobalSnapshotConsensusFunctions,
+  GlobalSnapshotStateChannelEventsProcessor,
+  GlobalSnapshotTraverse
+}
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema._
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.{Amount, Balance}
+import org.tessellation.schema.block.DAGBlock
+import org.tessellation.schema.epoch.EpochProgress
+import org.tessellation.schema.height.{Height, SubHeight}
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.schema.transaction.{DAGTransaction, TransactionReference}
+import org.tessellation.sdk.config.AppEnvironment
+import org.tessellation.sdk.domain.transaction.{TransactionChainValidator, TransactionValidator}
+import org.tessellation.sdk.infrastructure.block.processing.{BlockAcceptanceLogic, BlockAcceptanceManager, BlockValidator}
+import org.tessellation.sdk.infrastructure.metrics.Metrics
+import org.tessellation.security.hash.{Hash, ProofsHash}
+import org.tessellation.security.hex.Hex
+import org.tessellation.security.key.ops.PublicKeyOps
+import org.tessellation.security.signature.{Signed, SignedValidator}
+import org.tessellation.security.{Hashed, KeyPairGenerator, SecurityProvider}
+import org.tessellation.shared.sharedKryoRegistrar
+import org.tessellation.syntax.sortedCollection._
+import org.tessellation.tools.TransactionGenerator._
+import org.tessellation.tools.{DAGBlockGenerator, TransactionGenerator}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.{NonNegLong, PosInt}
+import org.scalacheck.Gen
+import weaver._
+import weaver.scalacheck.Checkers
+
+object GlobalSnapshotTraverseSuite extends MutableIOSuite with Checkers {
+  type GenKeyPairFn = () => KeyPair
+
+  type Res = (KryoSerializer[IO], SecurityProvider[IO], Metrics[IO], Random[IO], GenKeyPairFn)
+
+  override def sharedResource: Resource[IO, Res] = for {
+    kryo <- KryoSerializer.forAsync[IO](sharedKryoRegistrar)
+    implicit0(sp: SecurityProvider[IO]) <- SecurityProvider.forAsync[IO]
+    metrics <- Metrics.forAsync[IO](Seq.empty)
+    random <- Random.scalaUtilRandom[IO].asResource
+    mkKeyPair = () => KeyPairGenerator.makeKeyPair.unsafeRunSync()
```

Comment:
I need `mkKeyPair` to generate addresses in transaction generator. I don't know in advance how many keys I need , cause needed number is generated. Passing effect to `scalacheck.Gen` instead of impure function would be complicated.

Reply:
yeah, but you cannot create an arbitrary for keypair, so probably it can be done in a different way. for similar reasons you don't create a snapshot via Gen

Reply:
what was wrong with the existing implementation of TransactionGenerator?